### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by NazariyPogo**
Made node class.
Made tests for node class.
Unit tested node class.

### CODE REVIEW

**Comment by czamora7** [comment]
I will review this

File: GenericsHomework/GenericsHomeworkTests/GenericsHomeworkTests.csproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
```
Comment (czamora7) [inline GenericsHomework/GenericsHomeworkTests/GenericsHomeworkTests.csproj:3]: .NET analyzers should be enabled here

File: GenericsHomework/GenericsHomeworkTests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+namespace GenericsHomeworkTests
+{
+    [TestClass]
+    public class NodeTests
+    {
+        [TestMethod]
+        public void Node_ConstructorWorks()
+        {
+            Node<string> node = GetNode();
+            Assert.IsNotNull(node);
```
Comment (czamora7) [inline GenericsHomework/GenericsHomeworkTests/NodeTests.cs:14]: Assert.AreEqual<T> should be used for this assignment

**Review by czamora7 — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
Include a constuctor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ❌
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ❌
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ❌
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔

Fundamentals
Ensure you enable:
nullable reference types is enabled ✔
net6 targetted ✔
C# 10.0 ✔
and enabled .NET analyzers for both projects ❌
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: GenericsHomework/GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        public override string? ToString()
+        {
+            if(this.Value is null)
+            {
+                return null;
+            }
+
+            return this.Value.ToString();
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:66]:
```suggestion
return Value.ToString();
```
The above change will do the same thing.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            if (this.Value.Equals(value)) { return true; }
+
+            Node<T> tracer = this;
+
+            for (; tracer.Next != this; tracer = tracer.Next)
+            {
+                if(tracer.Value.Equals(value)) { return true; }
+            }
+
+            if (tracer.Value.Equals(value)) { return true; }
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:54]: You can also do this via a do-while loop, which may be cleaner to read.

File: GenericsHomework/GenericsHomeworkTests/NodeTests.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            Node<string> node = GetNode();
+            node.Append("Princess Buttercup");
+            Assert.AreEqual<string>("Inigo Montoya", node.Value);
+        }
+
+        [TestMethod]
+        public void Node_Append_NewValueAppended()
+        {
+            Node<string> node = GetNode();
+            node.Append("Princess Buttercup");
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomeworkTests/NodeTests.cs:43]: You should write some tests for inserting multiple values, inserting into the 'middle' (when i say middle, I mean change your node to not be the first one created) of the list, etc.

**Comment by mmwoodfo** [comment]
This coebase looks similar to #101

**Pull request closed without merging**